### PR TITLE
fixes shadow hydration escaping

### LINKED ISSUE #3773: [bug?] Shadow Endpoints fails to hydrate big data sent to their matching page because it is truncated. causing SyntaxError: Unexpected end of JSON input
### Describe the bug

i originally came to this issue when trying to use shadow endpoints for the first time:

image

EDIT: there are more discovery points in subsequent comments below.

### Reproduction

https://stackblitz.com/edit/sveltejs-kit-template-default-ahljwi?file=src/routes/about.svelte

`npm run dev`

and go to the /about endpoint

### Logs

```shell
9:03:33 PM [vite] hmr update /src/routes/about.svelte
```

### System Info

this is my mac envinfo since stackblitz doesnt seem to be able to run envinfo

```shell
  System:
    OS: macOS 11.4
    CPU: (8) x64 Apple M1
    Memory: 51.47 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.4.1 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.18.1 - /usr/local/bin/npm
  Browsers:
    Chrome: 98.0.4758.80
    Safari: 14.1.1
  npmPackages:
    @sveltejs/adapter-netlify: ^1.0.0-next.44 => 1.0.0-next.44 
    @sveltejs/kit: ^1.0.0-next.260 => 1.0.0-next.260 
    svelte: ^3.44.3 => 3.46.4 
```

### Severity

blocking an upgrade

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by PH4NTOMiki**
### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

fixes #3773
there are many better ways to solve this, if anyone has a suggestion.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by PH4NTOMiki** [comment]
I also tried using devalue package but it's the same, html becomes malformed. This way is vulnerable to XSS probably so not the best way, but just a suggestion.

**Comment by PH4NTOMiki** [comment]
Actually I found out that Next.js is using the same(similar) strategy with regex like this PR:
https://github.com/vercel/next.js/blob/694407450638b037673c6d714bfe4126aeded740/packages/next/server/htmlescape.ts

I'll improve this a little but this isn't that bad.

**Comment by dominikg** [comment]
we have a utility for that https://github.com/sveltejs/kit/blob/59badb515755a9294365e8084f9beefa10a90773/packages/kit/src/utils/escape.js#L19

**Comment by dominikg** [comment]
and it would be great to have a test in https://github.com/sveltejs/kit/tree/master/packages/kit/test/apps/basics/src/routes/xss  so that we can catch regressions.

**Comment by PH4NTOMiki** [comment]
> 

Thanks, I actually adopted the code from Next.js in the latest commit, but I'll implement that code now you suggested because DRY.

**Comment by dominikg** [comment]
@PH4NTOMiki  i pushed a fix for the invalid double-qoute escape and added a test. hope you don't mind. this is kind of urgent :grimacing:

**Comment by PH4NTOMiki** [comment]
> @PH4NTOMiki i pushed a fix for the invalid double-qoute escape and added a test. hope you don't mind. this is kind of urgent grimacing

Yeah, of course, thank you

File: packages/kit/test/apps/basics/test/test.js [hunk 1]
```diff
@@ -1992,4 +1992,14 @@ test.describe.parallel('XSS', () => {
 		// @ts-expect-error - check global injected variable
 		expect(await page.evaluate(() => window.pwned)).toBeUndefined();
 	});
+
+	test('no xss via shadow endpoint', async ({ page }) => {
+		await page.goto('/xss/shadow')
```
Comment (benmccann) [inline packages/kit/test/apps/basics/test/test.js:1997]:
```suggestion
		await page.goto('/xss/shadow');
```

**Comment by benmccann** [comment]
Looks like you'll need to run `pnpm format`

File: packages/kit/src/utils/escape.js [hunk 2]
```diff
@@ -15,7 +14,34 @@ const escape_json_string_in_html_dict = {
...
+	'"': '\\"',
+	...escape_json_in_html_dict
+};
+
+/**
+ * escape a json string to be embedded into a script data tag
+ *
+ * <script>
+ *   output here
+ * </script>
```
Comment (benmccann) [inline packages/kit/src/utils/escape.js:28]:
```suggestion
 * Escape a JSON string to be embedded into a `<script>` tag
```

File: packages/kit/src/utils/escape.js [hunk 3]
```diff
@@ -15,7 +14,34 @@ const escape_json_string_in_html_dict = {
...
+}
+
+/**
+ * escape a json string to be embedded into a larger json object thats going to be embedded in html
+ *
+ * <script>
+ *   {
+ *     "foo":"output here"
+ *   }
+ * </script>
```
Comment (benmccann) [inline packages/kit/src/utils/escape.js:42]:
```suggestion
 * Escape a stringified JSON object that's going to be embedded in a `<script>` tag
```

**Review by benmccann — COMMENTED** [review]
I had a couple suggestions for the jsdocs, but otherwise lgtm

I would like to get another set of eyes from @Conduitry though since he was more involved in https://github.com/sveltejs/kit/pull/2534

**Comment by dominikg** [comment]
these docs describe where the escaped result can be safely used, not how the return value looks like.

see https://github.com/sveltejs/kit/blob/647131e00149cb5299042eedeb880efcd855a2df/packages/kit/src/runtime/server/page/load_node.js#L260

it's actually a string prop on a larger json thats added to a script block elsewhere later. Not fond of this complexity but it is what it is.

I'd like to merge this first and we can improve the docs in a followup PR.

File: packages/kit/src/utils/escape.js [hunk 4]
```diff
@@ -15,7 +14,24 @@ const escape_json_string_in_html_dict = {
...
+ * @param {string} str
+ */
+export function escape_json_in_html(str) {
+	return escape(str, escape_json_in_html_dict, (code) => `\\u${code.toString(16).toUpperCase()}`);
+}
+
+/**
+ * Escape a stringified JSON object that's going to be embedded in a `<script>` tag
+ * @param {string} str
+ */
```
Comment (benmccann) [inline packages/kit/src/utils/escape.js:34]: oh wait. I think my documentation suggestions were incorrect. this should be better
```suggestion
 * Escape a stringified JSON object that's going to be embedded in a `<script>` tag
 * @param {string} str
 */
export function escape_json_in_html(str) {
	return escape(str, escape_json_in_html_dict, (code) => `\\u${code.toString(16).toUpperCase()}`);
}

/**
 * Escape a string JSON value to be embedded into a `<script>` tag
 * @param {string} str
 */
```

**Comment by benmccann** [comment]
Something in this PR seems to have broken https://kit.svelte.dev/docs

**Comment by PH4NTOMiki** [comment]
@dominikg what do you think?
@benmccann could you roll over back to https://github.com/sveltejs/kit/pull/3793/commits/886ad462f5881694c2856d93f78b64d92029632b
I think that should work?

**Comment by PH4NTOMiki** [comment]
Or should I try to PR that again?

**Comment by PH4NTOMiki** [comment]
@benmccann also, is there any change that that Vite update could be causing anything? maybe together with this, or alone?

**Comment by PH4NTOMiki** [comment]
I think it's the `"` fix that @dominikg pushed, https://github.com/sveltejs/kit/pull/3793/commits/8c8e4e68b166a0afd323eb52c7d87c37c93d59a0 this commit, I think is the cause

**Comment by benmccann** [comment]
It works if I revert this line:
```
body += `<script type="application/json" data-type="svelte-props">${escape_json_in_html(s(shadow_props))}</script>`;
```

**Comment by PH4NTOMiki** [comment]
Isn't that the current/latest line?

**Comment by benmccann** [comment]
Yes, and it's broken

**Pull request merged**